File: web/src/main/java/no/nav/sbl/dialogarena/rest/mappers/BekreftelseMapper.java
package no.nav.sbl.dialogarena.rest.mappers;

import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse;

import java.util.List;
import java.util.Optional;

public class BekreftelseMapper {

    public static void setBekreftelse(JsonOkonomiopplysninger opplysninger, String type, Boolean verdi, String tittel) {
        final Optional<JsonOkonomibekreftelse> utbetaltBekreftelse = opplysninger.getBekreftelse().stream()
                .filter(bekreftelse -> bekreftelse.getType().equals(type)).findFirst();

        if (utbetaltBekreftelse.isPresent()){
            utbetaltBekreftelse.get().withKilde(JsonKilde.BRUKER).withVerdi(verdi);
        } else {
            List<JsonOkonomibekreftelse> bekreftelser = opplysninger.getBekreftelse();
            bekreftelser.add(new JsonOkonomibekreftelse()
                    .withKilde(JsonKilde.BRUKER)
                    .withType(type)
                    .withTittel(tittel)
                    .withVerdi(verdi));
        }
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/nymodell/producer/json/mappers/okonomi/BekreftelseMapper.kt
package no.nav.sosialhjelp.soknad.nymodell.producer.json.mappers.okonomi

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse
import no.nav.sosialhjelp.soknad.nymodell.domene.okonomi.Bekreftelse
import no.nav.sosialhjelp.soknad.nymodell.domene.okonomi.BekreftelseRepository
import no.nav.sosialhjelp.soknad.nymodell.producer.json.createChildrenIfNotExists
import no.nav.sosialhjelp.soknad.nymodell.producer.json.mappers.DomainToJsonMapper
import no.nav.sosialhjelp.soknad.nymodell.producer.json.mappers.okonomi.type.toSoknadJsonType
import org.springframework.stereotype.Component
import java.util.*

@Component
class BekreftelseMapper(
    private val bekreftelseRepository: BekreftelseRepository
) : DomainToJsonMapper {
    override fun mapDomainToJson(soknadId: UUID, json: JsonInternalSoknad) {
        json.createChildrenIfNotExists()
        json.soknad.data.okonomi.opplysninger
            .bekreftelse
            .addAll(
                bekreftelseRepository.findAllBySoknadId(soknadId)
                    .map { it.toJsonOkonomibekreftelse() }
            )
    }

    private fun Bekreftelse.toJsonOkonomibekreftelse(): JsonOkonomibekreftelse {
        return JsonOkonomibekreftelse()
            .withKilde(JsonKilde.BRUKER) // alltid bruker
            .withType(type?.toSoknadJsonType())
            .withTittel(tittel)
            .withVerdi(bekreftet)
            .withBekreftelsesDato(dato.toString())
    }
}
